fix: improve handling of failed dnf list command

esolitos opened the pull request:
## What

Updates the script to delete old artifacts in Cloudsmith:

- Add a `|| true` after the `dnf list` command to ensure that the script continues to execute even if the `dnf list` command returns non-zero status code.
- Adds the standard set of bash options at the beginning of the script.

## Why

- fixes an issue where if there are no artifacts to remove the Workflow would be marked as failed

## Additional info

Failed workflows: https://github.com/vespa-engine/vespa/actions/workflows/delete-old-versions-in-archive.yml?query=is%3Afailure+branch%3Amaster


---

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

esolitos committed:
fix: improve handling of failed dnf list command

The changes made in this commit fix an issue where the script would fail if the `dnf list` command returned an error.

The script now uses the `|| true` operator to ensure that the script continues to execute even if the `dnf list` command fails.

esolitos committed:
chore: add more info to curl calls

esolitos committed:
fix: separate VERSIONS_TO_DELETE and VESPA_VERSIONS process

esolitos committed:
chore: improve cleaner script output

esolitos committed:
fix: shellckeck fixes

hakonhall commented on .github/scripts/delete-old-cloudsmith-artifacts.sh:
```diff
@@ -1,42 +1,67 @@
+fi
 
+# Note: Allow the last Vespa 7 release to remain in the repo!
+VERSIONS_TO_DELETE=$(echo "${VESPA_VERSIONS}" | awk '/[0-9].*\.[0-9].*\.[0-9].*/{print $2}' | sort -V | grep -v "${LAST_VESPA_7_RELEASE}" | head -n -$MAX_NUMBER_OF_RELEASES)
```
Over-generalized regex - you probably mean `/[0-9]+\.[0-9]+\.[0-9]+/` ?

esolitos replied:
I just copied the existing regex and indeed I like your suggestion better. 

I'll tell you what.. Given we want to handle the 8+ only, I'll give you one even better:

```suggestion
VERSIONS_TO_DELETE=$(echo "${VESPA_VERSIONS}" | awk '/[8-9]\.[0-9]+\.[0-9]+/{print $2}' | sort -V | head -n -$MAX_NUMBER_OF_RELEASES)
```

hakonhall replied:
Sweet

hakonhall commented on .github/scripts/delete-old-cloudsmith-artifacts.sh:
```diff
@@ -1,42 +1,67 @@
-      curl -sSLf -X DELETE \
+    echo '\033[1;33mDeleting the following RPMs:\033[0m'
+    echo
+    cat "${RPMS_TO_DELETE}"
```
I see you sometimes use curly braces when referencing parameters and sometimes not (when either is semantically equivalent).  If you want my 2 cents I prefer to avoid curly braces as it is easier on the eyes and fingers.

In a somewhat similar vein, I prefer to always reference parameters inside double-quotes (unless that would be semantically wrong), possibly with the exception of integer variables :thinking: .

esolitos replied:
> I see you sometimes use curly braces when referencing parameters and sometimes not (when either is semantically equivalent).

The "sometimes not" is only where I might have forgotten or copied some code from somewhere and not noticed. 
My approach is to use it _always_, for:
- consistency and not need a context switch to quickly identify variables
- using them everywhere possible avoids the need to remember edge cases _(like $9 is ok and $10 isn't)_
- it is consistent with other language templates (python for example)


>  I prefer to always reference parameters inside double-quotes (unless that would be semantically wrong), possibly with the exception of integer variables

That's also my take, if you see around I tend do always quote any usage of variables. Similarly to the `${}`, where they are lacking I probably missed them (unless it was required for the specific usage of the var) or was some code taken from an existing implementation. 😅

hakonhall replied:
I think you're in the minority on always wanting to use curly braces.

esolitos replied:
My understanding was that it is best practice. 🙃 

I'll leave this here and tackle this discussion when we look into a linter/code-style as part of [VESPANG-511](https://vespaai.atlassian.net/browse/VESPANG-511).

hakonhall replied:
:+1:

esolitos committed:
fix: address review comments

- if dnf list fails, but still prints something to stdout, then this failure will be ignored as it is now
- Over-generalized regex
- consistent usage of curly braces on variable and quotes on strings

hakonhall commented on .github/scripts/delete-old-cloudsmith-artifacts.sh:
```diff
@@ -1,42 +1,74 @@
 
-# Allow the last Vespa 7 release to remain in the repo
-VERSIONS_TO_DELETE=$(dnf list -y --quiet --showduplicates --disablerepo='*' --enablerepo=vespa-open-source-rpms vespa | awk '/[0-9].*\.[0-9].*\.[0-9].*/{print $2}' | sort -V | grep -v "7.594.36" | head -n -$MAX_NUMBER_OF_RELEASES)
+dnf list -y --quiet --showduplicates --disablerepo='*' --enablerepo=vespa-open-source-rpms vespa >/tmp/dnf-stdout 2>/tmp/dnf-stderr || true
```
By testing for failure I was thinking of testing the exit status, so something like

```
if DNF_STDOUT=$(dnf list -y --quiet --showduplicates); then
  ...
else
  ... failure, be sure to exit with some message, whether something was written to stderr or not
fi
```

Usually, a failing command writes something to stderr, so it is equivalent.

esolitos replied:
I believe your solution is better because `dnf` does not seem to care about the `--quiet`  option and _only the first run_ it writes _"Importing GPG key 0x0D35DA7B"_ to stderr... 😒

esolitos replied:
I solved it by adding `dnf makecache` before, so that it's pre-cached. 🙃

esolitos committed:
feat(ci): add dnf makecache to ensure repo is up-to-date

This prevents the output of  "Importing GPG key 0x0D35DA7B" during the list.

esolitos committed:
shellcheck: SC2028 (info): echo may not expand escape sequences. Use printf.
